### PR TITLE
StereoPannerNode backend

### PR DESCRIPTION
**Pull request by PurpleHairEngineer**
Implemented the backend for StereoPannerNode in the media crate by creating a new node implementation using PannerNode as a guide according to the specification.

### CODE REVIEW

**Comment by PurpleHairEngineer** [comment]
@Manishearth Okay, I updated stereo_panner_node.rs to reflect the changes you suggested. I will make another commit once I have made the example.

**Comment by PurpleHairEngineer** [comment]
@Manishearth Okay, I added an example. As I said to the dev-servo email list, I am still having trouble running any examples on my computer so I am not able to test my example. But it seems to compile using "cargo build". Please let me know if I need to change anything else.

File: examples/stereo_panner.rs [hunk 1]
```diff
@@ -0,0 +1,62 @@
+extern crate servo_media;
+extern crate servo_media_auto;
+
```
Comment (Manishearth) [inline examples/stereo_panner.rs]: Name this file `stereo_panner.rs`. Also there will need to be an entry in `examples/Cargo.toml` so that it can be run

Comment (Manishearth) [reply examples/stereo_panner.rs]: whoops, the name is already correct. you'll still need the entry though

File: audio/stereo_panner_node.rs [hunk 2]
```diff
@@ -0,0 +1,137 @@
+use block::{Block, Chunk, Tick, FRAMES_PER_BLOCK};
+use euclid::Vector3D;
```
Comment (Manishearth) [inline audio/stereo_panner_node.rs]: Please name this file `stereo_panner.rs`. Also, for this to work you'll need a `mod stereo_panner` in `audio/lib.rs`

File: audio/stereo_panner.rs [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        inputs
+    }
+
+    fn input_count(&self) -> u32 {
+        1
+    }
+
+    fn get_param(&mut self, id: ParamType) -> &mut Param {
+        match id {
+            ParamType::Position(Param::pan) => &mut self.pan,
```
Comment (Manishearth) [inline audio/stereo_panner.rs]: This needs to be `ParamType::Pan`

`param.rs` also needs a `Pan` variant added to `ParamType`

File: examples/stereo_panner.rs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+use servo_media::audio::node::{AudioNodeInit, AudioNodeMessage, AudioScheduledSourceNodeMessage};
+use servo_media::audio::param::{ParamType, RampKind, UserAutomationEvent};
+use servo_media::audio::stereo_panner_node::StereoPannerOptions;
```
Comment (Manishearth) [inline examples/stereo_panner.rs]: this import is broken now

File: audio/stereo_panner.rs [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+                    let mut pan = self.pan;
+
+                    // https://webaudio.github.io/web-audio-api/#stereopanner-algorithm
+
+                    // clamp pan to [-1, 1]
+                    pan = max(-1, pan);
+                    pan = min(1, pan);
+
+                    let x = if mono {
+                        x = (pan + 1) / 2;
```
Comment (Manishearth) [inline audio/stereo_panner.rs]: remove the `x=` and the semicolons in these three lines

File: audio/stereo_panner.rs [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+            } else {
+                debug_assert!(block.chan_count() == 2);
+                false
+            };
+
+                let (l, r) = block.data_mut().split_at_mut(FRAMES_PER_BLOCK.0 as usize);
+                for frame in 0..FRAMES_PER_BLOCK.0 {
+                    let frame = Tick(frame);
+                    self.update_parameters(info, frame);
+                    let mut pan = self.pan;
```
Comment (Manishearth) [inline audio/stereo_panner.rs]: `self.pan.value()`

This should actually be out of the loop, and within the loop you want something like

```rust
                if self.update_parameters(info, frame) {
                    param = self.param.value();
                }
```

File: audio/stereo_panner.rs [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+impl Default for StereoPannerOptions {
+    fn default() -> Self {
+        StereoPannerOptions {
+            pan: 0.,
+        }
+    }
+}
+
+#[derive(AudioNodeCommon)]
+pub(crate) struct StereoPannerNode {
```
Comment (Manishearth) [inline audio/stereo_panner.rs]: This needs a `channel_info` field. See the gain node code

File: audio/stereo_panner.rs [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+    pan: Param,
+}
+
+impl StereoPannerNode {
+    pub fn new(options: StereoPannerOptions, channel_info: ChannelInfo) -> Self {
+        Self {
+            pan: options.pan,
```
Comment (Manishearth) [inline audio/stereo_panner.rs]: `Param::new(options.pan)`

File: audio/stereo_panner.rs [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+    }
+
+    fn get_param(&mut self, id: ParamType) -> &mut Param {
+        match id {
+            ParamType::Position(Param::pan) => &mut self.pan,
+            _ => panic!("Unknown param {:?} for PannerNode", id),
+        }
+    }
+
+    fn message_specific(&mut self, message: AudioNodeMessage, _sample_rate: f32) {
```
Comment (Manishearth) [inline audio/stereo_panner.rs]: This function isn't necessary, the `get_param` deals with all of this. You only need this if you have non-Param "parameters"

File: audio/stereo_panner.rs [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+
+// .normalize(), but it takes into account zero vectors
+use panner_node::normalize_zero;
+
+#[derive(Copy, Clone, Debug)]
+pub struct StereoPannerOptions {
+    pub pan: f32,
+}
+
+pub enum StereoPannerNodeMessage {
```
Comment (Manishearth) [inline audio/stereo_panner.rs]: Not necessary

File: audio/stereo_panner.rs [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+                    // https://webaudio.github.io/web-audio-api/#stereopanner-algorithm
+
+                    // clamp pan to [-1, 1]
+                    pan = max(-1, pan);
+                    pan = min(1, pan);
+
+                    let x = if mono {
+                        x = (pan + 1) / 2;
+                    } else if pan <= 0. {
+                        x = pan + 1;
```
Comment (Manishearth) [inline audio/stereo_panner.rs]: almost all of the numerical values here need to have a decimal point after them to make them into floats

File: audio/stereo_panner.rs [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+                let (l, r) = block.data_mut().split_at_mut(FRAMES_PER_BLOCK.0 as usize);
+                for frame in 0..FRAMES_PER_BLOCK.0 {
+                    let frame = Tick(frame);
+                    self.update_parameters(info, frame);
+                    let mut pan = self.pan;
+
+                    // https://webaudio.github.io/web-audio-api/#stereopanner-algorithm
+
+                    // clamp pan to [-1, 1]
+                    pan = max(-1, pan);
```
Comment (Manishearth) [inline audio/stereo_panner.rs]: floats aren't `Ord`, you'll have to do this with:

```rust
                    pan = if pan < -1. {
                        -1.
                    } else if pan > 1. {
                        1.
                    } else {
                        pan
                    };
```

File: audio/stereo_panner.rs [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+    }
+}
+
+#[derive(AudioNodeCommon)]
+pub(crate) struct StereoPannerNode {
+    pan: Param,
+}
+
+impl StereoPannerNode {
+    pub fn new(options: StereoPannerOptions, channel_info: ChannelInfo) -> Self {
```
Comment (Manishearth) [inline audio/stereo_panner.rs]: You'll need to update `AudioNodeInit` to have `StereoPannerOptions` inside its `StereoPannerNode` variant

File: audio/stereo_panner.rs [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+
+impl AudioNodeEngine for StereoPannerNode {
+    fn node_type(&self) -> AudioNodeType {
+        AudioNodeType::StereoPannerNode
+    }
+
+    fn process(&mut self, mut inputs: Chunk, info: &BlockInfo) -> Chunk {
+        debug_assert!(inputs.len() == 1);
+
+        return inputs;
```
Comment (jdm) [inline audio/stereo_panner.rs]: Just pointing out that the rust compiler should be complaining about a lot of unreachable code after this line.

File: audio/stereo_panner.rs [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+            } else {
+                debug_assert!(block.chan_count() == 2);
+                false
+            };
+
+                let (l, r) = block.data_mut().split_at_mut(FRAMES_PER_BLOCK.0 as usize);
+                for frame in 0..FRAMES_PER_BLOCK.0 {
+                    let frame = Tick(frame);
+                    if self.update_parameters(info, frame) {
+                    pan = self.pan.value();
```
Comment (Manishearth) [inline audio/stereo_panner.rs]: You need to initialize this outside of the loop with `let pan = self.pan.value();`

Also, please run `rustfmt` on this file.

File: audio/stereo_panner.rs [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+#[derive(AudioNodeCommon)]
+pub(crate) struct StereoPannerNode {
+    channel_info: ChannelInfo,
+    pan: Param,
+}
+
+impl StereoPannerNode {
+    pub fn new(options: StereoPannerOptions, channel_info: ChannelInfo) -> Self {
+        Self {
+            Param::new(options.pan),
```
Comment (Manishearth) [inline audio/stereo_panner.rs]: `pan: Param::new(....)`

**Comment by PurpleHairEngineer** [comment]
@Manishearth Thank you. I've just been committing incrementally as I've been fixing the problems you mention. I'm sorry if you get a notification every time. It seems to build now so please take a look.

**Comment by Manishearth** [comment]
Oh, I see, thanks.

I'll review in a bit.

File: audio/stereo_panner.rs [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+            let mono = if block.chan_count() == 1 {
+                block.resize_silence(2);
+                true
+            } else {
+                debug_assert!(block.chan_count() == 2);
+                false
+            };
+
+            let (l, r) = block.data_mut().split_at_mut(FRAMES_PER_BLOCK.0 as usize);
+            let pan = self.pan.value();
```
Comment (Manishearth) [inline audio/stereo_panner.rs]: needs to be mutable

File: audio/node.rs [hunk 18]
```diff
@@ -28,7 +29,7 @@ pub enum AudioNodeInit {
     PannerNode(PannerNodeOptions),
     PeriodicWave,
     ScriptProcessorNode,
-    StereoPannerNode,
+    StereoPannerNode(StereoPannerOptions),
```
Comment (Manishearth) [inline audio/node.rs]: Also needs to be handled in `create_node` in `render_thread`

File: audio/stereo_panner.rs (hunk 14 above)
Comment (Manishearth) [reply audio/stereo_panner.rs]: This line still needs to be removed

**Comment by PurpleHairEngineer** [comment]
@Manishearth I made the changes you suggested, but it looks like all the current warnings are either because a variable was never used or because a variable does not need to be mutable. It looks like a lot of the warnings are outside the scope of the files I've changed.

**Comment by PurpleHairEngineer** [comment]
@Manishearth It seems to build successfully now, and all the warnings are for files/parts unmodified by the committed code. Could you please take another look?

File: examples/stereo_panner.rs [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+    context.connect_ports(osc.output(0), pan.input(0));
+    context.connect_ports(pan.output(0), dest.input(0));
+    let _ = context.resume();
+    context.message_node(
+        osc,
+        AudioNodeMessage::AudioScheduledSourceNode(AudioScheduledSourceNodeMessage::Start(0.)),
+    );
+    // 2s: Set pan to -1
+    context.message_node(
+        osc,
```
Comment (Manishearth) [inline examples/stereo_panner.rs]: This needs to be `pan`

```suggestion
        pan,
```

**Comment by PurpleHairEngineer** [comment]
@Manishearth I am trying to squash my commits based on the instructions that were given in https://github.com/servo/media/pull/242 . However, when I run `git rebase -i HEAD~15`, I see a lot of additional commits from before I started committing. I am not sure how to handle this.

**Comment by Manishearth** [comment]
It's because you merged master into this PR instead of rebasing over master. No big deal, I'll fix it.

**Comment by PurpleHairEngineer** [comment]
@Manishearth Sorry! So I should squash them all? Or just do nothing? I ran `rustfmt` on the modified files like you asked.

**Comment by Manishearth** [comment]
Actually it's just the first merge commit, remove everything before "Implemented the backend for StereoPannerNode" in the rebase dialog before using `fixup` on everything after that commit

**Comment by PurpleHairEngineer** [comment]
@Manishearth Okay, I did that. It stopped midway after I ran the `fixup` command and told me I had a merge conflict in Cargo.toml, which I thought I resolved. It seems to have squashed the commits, but now it says there are conflicts that must be resolved. I'm not sure how to fix this?

**Comment by Manishearth** [comment]
Can you tick the "allow collaborators to push to this pull request" checkbox on the side of this PR? I've got it rebased.

**Comment by PurpleHairEngineer** [comment]
@Manishearth The only button I see is "Allow edits from maintainers", and it was already checked, but I just unchecked it and checked it again.

**Comment by Manishearth** [comment]
Ah, i had the fork wrong. Fixed.

**Comment by Manishearth** [comment]
@bors-servo r+

Thanks!

**Pull request merged**